python/ray/util/sgd/v2/backends/backend.py

```diff
@@ -55,20 +59,140 @@ def start(self, initialization_hook: Optional[Callable[[], None]] = None):
             self.worker_group.execute(initialization_hook)
         self._backend.on_start(self.worker_group, self._backend_config)
 
-    def run(self, train_func: Callable[[], T]) -> List[T]:
-        """Executes a training function on all workers.
+    def start_training(self, train_func: Callable[[], T]) -> None:
+        """Executes a training function on all workers in a separate thread.
 
         Args:
             train_func (Callable): The training function to run on each worker.
+        """
+
+        # First initialize the session.
+        def initialize_session(world_rank):
+            thread = PropagatingThread(target=train_func, daemon=True)
+            try:
+                init_session(training_thread=thread, world_rank=world_rank)
+            except ValueError:
+                raise RuntimeError("Attempting to start training but a "
+                                   "previous training run is still ongoing. "
+                                   "You must call `finish_training` before "
+                                   "calling `start_training` again.") from None
+
+        futures = []
+        for world_rank in range(len(self.worker_group)):
+            futures.append(
+                self.worker_group.execute_single_async(
+                    world_rank, initialize_session, world_rank=world_rank))
+
+        ray.get(futures)
+
+        # Run the training function asynchronously in its own thread.
+        def train_async():
+            session = get_session()
+            session.training_thread.start()
+
+        self.worker_group.execute_async(train_async)
+
+    def fetch_next_result(self) -> Optional[List[Dict]]:
```

Comment:
How about renaming this to `fetch_next_report` to differentiate between what is returned by `train_func` and what is reported by `sgd.report()`? Same comment goes for other places we use `result` when referring to information sent by `sgd.report()`

Reply:
As discussed over Slack, sticking with result here.